src/projection.jl

```diff
@@ -1,24 +1,29 @@
 """
-    ProjectTo(x::T)
+    (p::ProjectTo{T})(dx)
 
-Returns a `ProjectTo{T,...}` functor able to project a differential `dx` onto the same tangent space as `x`.
-This functor encloses over what ever is needed to be able to be able to do that projection.
-For example, when projecting `dx=ZeroTangent()` on an array `P=Array{T, N}`, the size of `x`
-is not available from `P`, so it is stored in the functor.
+Projects the differential `dx` onto a specific cotangent space.
+This guaranees `p(dx)::T`, except for allowing `dx::AbstractZero` to pass through.
 
-    (::ProjectTo{T})(dx)
+Usually `T` is the "outermost" part of the type, and `p` stores additional 
+properties such as projectors for each constituent field,
+and a projector `p.element` for the element type of an array of numbers.
+These properties can be supplied as keyword arguments on construction,
+`p = ProjectTo{T}(; field=data, element=Projector(x))`. For each `T` in use,
+corresponding methods should be written for `ProjectTo{T}(dx)` with nonzero `dx`.
```

Comment:
Yeah i like this much better.

Should we also comment that `T` is `AbstractArray` for any general matrixes that do not have special structure?

Reply:
I put an example to this effect in the other docstring, https://github.com/JuliaDiff/ChainRulesCore.jl/pull/391/files#diff-39e64f3e0b5b507d448f9fe9bb7e31d6eb20927afb9610dd3a6f89d4b04f2e27R106

I haven't written up specifically the rules for what numbers are Real vs Float32 etc. either. But the examples capture a fair bit of it.